lib/standard/bytes.nit

```diff
@@ -0,0 +1,151 @@
+# limitations under the License.
+
+# Services for byte streams and arrays
+module bytes
+
+import kernel
+import collection::array
+intrude import string
+
+# A buffer containing Byte-manipulation facilities
+#
+# TODO: Change the bound to Byte when available in stdlib and bootstrap
+class ByteBuffer
+	super AbstractArray[Int]
+
+	# A NativeString being a char*, it can be used as underlying representation here.
+	private var items: NativeString
+
+	# Number of bytes in the array
+	redef var length: Int
+
+	# Capacity of the array
+	private var capacity: Int
+
+	# Has this buffer been persisted (to_s'd)?
+	private var persisted = false
+
+	init empty do
+		var ns = new NativeString(0)
+		init(ns, 0, 0)
+	end
+
+	init with_capacity(cap: Int) do
+		var ns = new NativeString(cap)
+		init(ns, 0, cap)
+	end
+
+	redef fun is_empty do return length != 0
+
+	redef fun [](i: Int): Int do
+		assert i >= 0
+		assert i < length
+		return items[i].ascii
+	end
+
+	redef fun []=(i: Int, v: Int) do
+		if persisted then regen
+		assert i >= 0
+		assert i < length
+		items[i] = v.ascii
+	end
+
+	redef fun add(c: Int) do
+		if persisted then regen
+		if length >= capacity then
+			enlarge(length)
+		end
+		items[length] = c.ascii
+		length += 1
+	end
+
+	redef fun append(arr: Collection[Int]) do
+		if arr isa ByteBuffer then
+			append_ns(arr.items, arr.length)
+		else
+			for i in arr do add i
+		end
+	end
+
+	redef fun clear do length = 0
+
+	# Regenerates the buffer, necessary when it was persisted
+	private fun regen do
+		var nns = new NativeString(capacity)
+		items.copy_to(nns, length, 0, 0)
+		persisted = false
+	end
+
+	# Appends the `ln` first bytes of `ns` to self
+	fun append_ns(ns: NativeString, ln: Int) do
+		if persisted then regen
+		var nlen = length + ln
+		if nlen > capacity then enlarge(nlen)
+		ns.copy_to(items, ln, 0, length)
+		length += ln
+	end
+
+	# Appends `ln` bytes from `ns` starting at index `from` to self
+	fun append_ns_from(ns: NativeString, ln, from: Int) do
+		if persisted then regen
+		var nlen = length + ln
+		if nlen > capacity then enlarge(nlen)
+		ns.copy_to(items, ln, from, length)
+		length += ln
+	end
+
+	redef fun enlarge(sz) do
+		if capacity >= sz then return
+		persisted = false
+		while capacity < sz do capacity = capacity * 2 + 2
+		var ns = new NativeString(capacity)
+		items.copy_to(ns, length, 0, 0)
+		items = ns
+	end
+
+	redef fun to_s do
+		persisted = true
+		return new FlatString.with_infos(items, length, 0, length -1)
+	end
+
+	redef fun iterator do return new BytesIterator.with_buffer(self)
+end
+
+private class BytesIterator
+	super IndexedIterator[Int]
+
+	var tgt: NativeString
+
+	var pos: Int
+
+	var max: Int
+
+	init with_buffer(b: ByteBuffer) do init(b.items, 0, b.length - 1)
+
+	redef fun is_ok do return pos < max
+
+	redef fun next do pos += 1
+
+	redef fun item do return tgt[pos].ascii
+
+	redef fun index do return pos
+end
+
+redef class NativeString
+	fun to_buffer: ByteBuffer do
```

Comment:
since `to_s` gives a `String` I could have expected that `to_buffer` gives a `Buffer`.
Maybe `to_bytes`?